test/test_graph.py

```diff
@@ -304,93 +316,180 @@ def _make_dp(self):
+        single_br_dp, _, _, ch1, ch2, *_, cir_map_dp, _, graph = self._make_dp()
         graph = replace_by_dummy(graph, ch1)
-        dps = find_replicable_branches(graph)
+        dps = find_non_dispatching_branches(graph)
         self.assertEqual(len(dps), 3)
         self.assertTrue(all(dp in (single_br_dp, ch2, cir_map_dp) for dp in dps))
 
         single_br_dp, *_, fork_zip_dp, _, cir_map_dp, _, graph = self._make_dp()
         graph = replace_by_dummy(graph, cir_map_dp)
-        dps = find_replicable_branches(graph)
+        dps = find_non_dispatching_branches(graph)
         self.assertTrue(all(dp in (single_br_dp, fork_zip_dp) for dp in dps))
 
         *_, end_dp, graph = self._make_dp()
         graph = replace_by_dummy(graph, end_dp)
-        dps = find_replicable_branches(graph)
+        dps = find_non_dispatching_branches(graph)
         self.assertEqual(len(dps), 0)
 
         single_br_dp, *_, fork_zip_dp, _, cir_map_dp, _, graph = self._make_dp()
         graph = replace_by_dummy(graph, fork_zip_dp)
-        dps = find_replicable_branches(graph)
+        dps = find_non_dispatching_branches(graph)
         self.assertEqual(len(dps), 2)
         self.assertTrue(all(dp in (single_br_dp, cir_map_dp) for dp in dps))
 
+    def test_single_non_replicable_dp(self):
+        # All replicable
+        *_, end_dp, graph = self._make_dp()
+        dps = _find_replicable_branches(graph)
+        self.assertEqual(len(dps), 1)
+        self.assertEqual(dps[0], end_dp)
+
+        single_br_dp, *_, fork_zip_dp, _, cir_map_dp, _, graph = self._make_dp()
+        make_non_replicable_dp(single_br_dp)
+        dps = _find_replicable_branches(graph)
+        self.assertEqual(len(dps), 2)
+        self.assertTrue(all(dp in (fork_zip_dp, cir_map_dp) for dp in dps))
+
+        single_br_dp, *_, ch1, ch2, fork_zip_dp, _, cir_map_dp, _, graph = self._make_dp()
+        make_non_replicable_dp(fork_zip_dp)
+        dps = _find_replicable_branches(graph)
+        self.assertEqual(len(dps), 4)
+        self.assertTrue(all(dp in (single_br_dp, ch1, ch2, cir_map_dp) for dp in dps))
+
+        single_br_dp, _, forker_dp, ch1, *_, cir_map_dp, _, graph = self._make_dp()
+        make_non_replicable_dp(ch1)
+        dps = _find_replicable_branches(graph)
+        self.assertEqual(len(dps), 3)
+        self.assertTrue(all(dp in (single_br_dp, forker_dp, cir_map_dp) for dp in dps))
+
+        single_br_dp, *_, fork_zip_dp, cir_br_dp, cir_map_dp, _, graph = self._make_dp()
+        make_non_replicable_dp(cir_map_dp)
+        dps = _find_replicable_branches(graph)
+        self.assertEqual(len(dps), 3)
+        self.assertTrue(all(dp in (single_br_dp, fork_zip_dp, cir_br_dp) for dp in dps))
+
+        single_br_dp, *_, fork_zip_dp, _, cir_map_dp, end_dp, graph = self._make_dp()
+        make_non_replicable_dp(end_dp)
+        dps = _find_replicable_branches(graph)
+        self.assertEqual(len(dps), 3)
```

Comment:
I assume we don't support execution of this case yet? (e.g. there are 3 branches for replicable dp).

Not for this PR. -- can we have a test case simply test if we can return 1 branch of replicable dp (but it's not the end_dp). Essentially simulate the production use case (e.g. after all dp there is a full sync dp)

Reply:
> I assume we don't support execution of this case yet? (e.g. there are 3 branches for replicable dp).

No, we don't. Those tests aim to test the graph function.

> can we have a test case simply test if we can return 1 branch of replicable dp (but it's not the end_dp). Essentially simulate the production use case (e.g. after all dp there is a full sync dp)

I will add one shortly.